src/ChannelManager.cpp

```diff
@@ -0,0 +1,210 @@
+// ChannelManager.cpp
+
+// Implements the cChannelManager class which stores and calls channel callbacks
+
+#include "Globals.h"  // NOTE: MSVC stupidness requires this to be the same across all modules
+
+#include <algorithm>
+
+#include "ChannelManager.h"
+#include "ChannelCallback.h"
+#include "ClientHandle.h"
+#include "Entities/Player.h"
+#include "OSSupport/CriticalSection.h"
+#include "World.h"
+
+
+class cChannelBroadcastCallback : public cPlayerListCallback
+{
+	AString m_Channel;
+	AString m_Data;
+public:
+	cChannelBroadcastCallback(const AString & a_Channel, const AString & a_Data)
+	: m_Channel(a_Channel), m_Data(a_Data){}
+
+	virtual bool Item(cPlayer * a_Type)
+	{
+		cClientHandle * Handle = a_Type->GetClientHandle();
+		if (Handle->HasPluginChannel(m_Channel))
+		{
+			Handle->SendPluginMessage(m_Channel, m_Data);
+		}
+		return false;
+	}
+};
+
+
+
+
+
+bool cChannelManager::RegisterChannel(const AString & a_Channel, cChannelCallbackPtr a_Callback)
+{
+	cCSLock Lock(m_CSCallbacks);
+	if (m_ChannelCallbacks.find(a_Channel) == m_ChannelCallbacks.end())
+	{
+		m_ChannelCallbacks.insert(std::make_pair(a_Channel, a_Callback));
```

Comment:
I think (?) insert will tell you if the value was already present, so you can do a one-liner.

Also, if you use emplace, I again think (?) you can drop the make_pair.